ARCHITECTURE.md

```diff
@@ -0,0 +1,3386 @@
+# The Cortex: Architectural Document
+
+## Emergent Narrative System for AI-Driven RPG
+
+**Version:** 1.0
+**Date:** December 2024
+
+---
+
+## Table of Contents
+
+1. [General Concept and Design Philosophy](#part-1-general-concept-and-design-philosophy)
+2. [Architecture of Two Crates](#part-2-architecture-of-two-crates)
+3. [Interaction and Data Flow](#part-3-interaction-and-data-flow)
+4. [Final Recommendations](#part-4-final-recommendations)
+
+---
+
+## Part 1: General Concept and Design Philosophy
+
+### 1.1 Vision
+
+The Cortex is an autonomous core for managing dynamic narrative, memory, and AI interaction in RPGs. It serves as the "brain" of emergent storytelling - a system where narrative emerges naturally from the interplay of game state, accumulated knowledge, and AI-generated responses.
+
+### 1.2 Core Design Principles
+
+#### 1.2.1 State-Driven Architecture
+
+All AI actions must result from analysis of current world state and accumulated memory. The system never generates content in isolation - every narrative output is grounded in:
+
+- Current `WorldState` (character positions, time, weather, ongoing events)
+- Historical facts stored in the `KnowledgeBase`
+- Active relationships and their emotional states
+- Recent events that provide immediate context
+
+```rust
+/// The fundamental principle: AI decisions are pure functions of state
+fn generate_narrative(
+    world_state: &WorldState,
+    knowledge: &KnowledgeGraph,
+    event: &GameEvent,
+) -> NarrativeOutput {
+    // All inputs are explicit - no hidden state
+}
+```
+
+#### 1.2.2 Event-Driven Communication
+
+The core reacts to external events rather than controlling the game loop. This separation ensures:
+
+- **Loose coupling**: The game engine controls timing and flow
+- **Responsiveness**: Events are processed asynchronously
+- **Flexibility**: Any game engine can integrate with the core
+
+```rust
+/// Events flow INTO the core, responses flow OUT
+pub trait EventProcessor {
+    fn process_event(&mut self, event: GameEvent) -> Vec<NarrativeResponse>;
+}
+```
+
+#### 1.2.3 Extensibility Through Traits and Enums
+
+The architecture uses Rust's type system to enable extension without modification:
+
+```rust
+/// New fact types can be added by extending this enum
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub enum FactType {
+    Relationship(RelationshipFact),
+    Event(EventFact),
+    Secret(SecretFact),
+    Custom(Box<dyn CustomFact>),
```

Comment:
How thats possible to generate `dyn CustomFact` by ai? 
But maybe we should allow to `game engine` to generate facts